File: src/go/utf-256.go
// UTF-256 Encoder/Decoder in Go (no external packages)

package main

import (
	"errors"
	"flag"
	"fmt"
	"io"
	"os"
)

func printUsage() {
	fmt.Fprintf(os.Stderr, `
UTF-256 Encoder/Decoder

Usage:
  utf256 -e UTF-8_INPUT -o OUTPUT_FILE
  utf256 -d UTF-256_INPUT -o OUTPUT_FILE

`)
}

func encodeToUTF256(input []byte) []byte {
	output := make([]byte, 0, len(input)*8)

	for _, b := range input {
		for i := 7; i >= 0; i-- {
			if (b>>i)&1 == 1 {
				output = append(output, 0xFF)
			} else {
				output = append(output, 0x00)
			}
		}
	}

	return output
}

func decodeFromUTF256(input []byte) ([]byte, error) {
	if len(input)%8 != 0 {
		return nil, errors.New("input is not a multiple of 8 bytes")
	}

	output := make([]byte, 0, len(input)/8)

	for i := 0; i < len(input); i += 8 {
		var b byte
		for j := 0; j < 8; j++ {
			b <<= 1
			switch input[i+j] {
			case 0x00:
				// nothing
			case 0xFF:
				b |= 1
			default:
				return nil, fmt.Errorf("invalid byte in UTF-256 stream: 0x%02X", input[i+j])
			}
		}
		output = append(output, b)
	}

	return output, nil
}

func readFile(path string) ([]byte, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	return io.ReadAll(file)
}

func writeFile(path string, data []byte) error {
	file, err := os.Create(path)
	if err != nil {
		return err
	}
	defer file.Close()

	_, err = file.Write(data)
	return err
}

func main() {
	var encodePath string
	var decodePath string
	var outputPath string

	flag.StringVar(&encodePath, "e", "", "Input UTF-8 file to encode")
	flag.StringVar(&decodePath, "d", "", "Input UTF-256 file to decode")
	flag.StringVar(&outputPath, "o", "", "Output file")
	flag.Parse()

	// Enforce mutual exclusivity and required output
	if (encodePath == "" && decodePath == "") || (encodePath != "" && decodePath != "") {
		fmt.Fprintln(os.Stderr, "Error: You must use exactly one of -e or -d.\n")
		printUsage()
		os.Exit(1)
	}

	if outputPath == "" {
		fmt.Fprintln(os.Stderr, "Error: Output file (-o) is required.\n")
		printUsage()
		os.Exit(1)
	}

	var inputData []byte
	var err error

	if encodePath != "" {
		inputData, err = readFile(encodePath)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error reading input file: %v\n", err)
			os.Exit(1)
		}

		encoded := encodeToUTF256(inputData)
		err = writeFile(outputPath, encoded)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error writing output file: %v\n", err)
			os.Exit(1)
		}

		fmt.Printf("Encoded to '%s'\n", outputPath)

	} else {
		inputData, err = readFile(decodePath)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error reading input file: %v\n", err)
			os.Exit(1)
		}

		decoded, err := decodeFromUTF256(inputData)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Decoding failed: %v\n", err)
			os.Exit(1)
		}

		err = writeFile(outputPath, decoded)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error writing output file: %v\n", err)
			os.Exit(1)
		}

		fmt.Printf("Decoded to '%s'\n", outputPath)
	}
}

File: src/python/utf-256.py
#!/usr/bin/env python3

import argparse
import os
import sys

def utf_256_encoder(text: str) -> bytes:
    result = []
    utf8_bytes = text.encode('utf-8')
    for byte in utf8_bytes:
        for i in range(8):
            bit = (byte >> (7 - i)) & 1
            result.append(0xFF if bit else 0x00)
    return bytes(result)

def utf_256_decoder(data: bytes) -> str:
    if len(data) % 8 != 0:
        raise ValueError("Invalid UTF-256 data: length must be a multiple of 8")

    utf8_bytes = []
    for i in range(0, len(data), 8):
        byte = 0
        for j in range(8):
            b = data[i + j]
            if b == 0xFF:
                bit = 1
            elif b == 0x00:
                bit = 0
            else:
                raise ValueError(f"Invalid byte: {b}, must be 0x00 or 0xFF")
            byte = (byte << 1) | bit
        utf8_bytes.append(byte)

    return bytes(utf8_bytes).decode('utf-8')

def main():
    # Define the custom help message as a constant
    CUSTOM_HELP = """
------------------------------------------------------
UTF-256 encoder/decoder

./utf-256.py [FLAG] INPUT_FILE -o OUTPUT_FILE

|-----------|------------------------------------------|
|  FLAG     |  What it does                            |
|-----------|------------------------------------------|
| -e        | Encodes a UTF-8 file into a UTF-256 file |
| --encode  |                                          |
|-----------|------------------------------------------|
| -d        | Decodes a UTF-256 file into a UTF-8 file |
| --decode  |                                          |
|-----------|------------------------------------------|
"""

    # Check for help flag manually before parsing
    if "-h" in sys.argv or "--help" in sys.argv or len(sys.argv) == 1:
        print(CUSTOM_HELP)
        sys.exit(0)

    try:
        # Configure the parser without default help messages
        parser = argparse.ArgumentParser(add_help=False)

        group = parser.add_mutually_exclusive_group(required=True)
        group.add_argument("-e", "--encode")
        group.add_argument("-d", "--decode")
        parser.add_argument("-o", "--output", required=True)

        # Parse arguments inside a try block
        args = parser.parse_args()

        if args.encode:
            input_file = args.encode
            if not os.path.isfile(input_file):
                print(f"Error: UTF-8 input file '{input_file}' does not exist.")
                sys.exit(1)
            
            with open(input_file, 'rb') as f:
                utf8_bytes = f.read()
            text = utf8_bytes.decode('utf-8')
            encoded = utf_256_encoder(text)
            
            with open(args.output, 'wb') as f:
                f.write(encoded)
            print(f"Encoded successfully to '{args.output}'")
        
        elif args.decode:
            input_file = args.decode
            if not os.path.isfile(input_file):
                print(f"Error: UTF-256 input file '{input_file}' does not exist.")
                sys.exit(1)
            
            with open(input_file, 'rb') as f:
                data = f.read()
            decoded_bytes = utf_256_decoder(data).encode('utf-8')
            
            with open(args.output, 'wb') as f:
                f.write(decoded_bytes)
            print(f"Decoded successfully to '{args.output}'")

    except SystemExit:
        # This catches errors like missing required arguments
        print(CUSTOM_HELP)
        sys.exit(1)
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()
